### PR TITLE
empty cluster status when cluster is offline

### PR DESCRIPTION
**Pull request by zhiweiyin318**
related issue: https://github.com/open-cluster-management/backlog/issues/13596

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -29,3 +31,12 @@ func BuildKubeClient(kubeConfigPath string) (*kubernetes.Clientset, error) {
 func ResourceNamespacedName(resourceType, namespace, name string) string {
 	return fmt.Sprintf("%s/%s/%s", resourceType, namespace, name)
 }
+
+func ClusterIsOffLine(conditions []metav1.Condition) bool {
+	for _, sc := range conditions {
```
Comment (qiujian16) [inline pkg/utils/utils.go:36]: use meta.IsConditionTrue

File: pkg/controllers/clusterinfo/clusterinfo_controller.go [hunk 2]
```diff
@@ -188,15 +188,21 @@ func (r *ClusterInfReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	// TODO: the conditions of managed cluster need to be deprecated.
-	newConditions := cluster.Status.Conditions
+	clusterConditions := cluster.Status.Conditions
+	newClusterInfo := clusterInfo.DeepCopy()
+
+	// empty clusterInfo status when cluster is offline
+	if utils.ClusterIsOffLine(clusterConditions) {
+		newClusterInfo.Status = clusterinfov1beta1.ClusterInfoStatus{}
```
Comment (qiujian16) [inline pkg/controllers/clusterinfo/clusterinfo_controller.go:196]: I think we only empty the node list here.

File: pkg/utils/utils.go (hunk 1 above)
Comment (zhiweiyin318) [reply pkg/utils/utils.go]: done

File: pkg/controllers/clusterinfo/clusterinfo_controller.go (hunk 2 above)
Comment (zhiweiyin318) [reply pkg/controllers/clusterinfo/clusterinfo_controller.go]: done

File: pkg/klusterlet/nodecollector/nodecollecter.go [hunk 3]
```diff
@@ -108,6 +109,10 @@ func (r *resourceCollector) reconcile(ctx context.Context) {
 		klog.Errorf("Failed to get cluster: %v", err)
 	}
 
+	if utils.ClusterIsOffLine(clusterinfo.Status.Conditions) {
```
Comment (qiujian16) [inline pkg/klusterlet/nodecollector/nodecollecter.go:112]: I do not think this is needed. Are you  saying cluster is offline but addon is  still running?

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -29,3 +32,7 @@ func BuildKubeClient(kubeConfigPath string) (*kubernetes.Clientset, error) {
 func ResourceNamespacedName(resourceType, namespace, name string) string {
 	return fmt.Sprintf("%s/%s/%s", resourceType, namespace, name)
 }
+
+func ClusterIsOffLine(conditions []metav1.Condition) bool {
+	return !meta.IsStatusConditionTrue(conditions, clusterapiv1.ManagedClusterConditionAvailable)
```
Comment (qiujian16) [inline pkg/utils/utils.go:37]: we should  only consider unknown  status

File: pkg/klusterlet/clusterinfo/clusterinfo_controller.go [hunk 5]
```diff
@@ -72,6 +73,10 @@ func (r *ClusterInfoReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	if utils.ClusterIsOffLine(clusterInfo.Status.Conditions) {
+		return ctrl.Result{RequeueAfter: time.Minute * 5}, nil
```
[the remark this answers is not in the document]
Comment (zhiweiyin318) [reply pkg/klusterlet/clusterinfo/clusterinfo_controller.go:77]: we do not watch the upgrade info like channel/version of OCP, need to requeue to check these info.

File: pkg/klusterlet/nodecollector/nodecollecter.go (hunk 3 above)
Comment (zhiweiyin318) [reply pkg/klusterlet/nodecollector/nodecollecter.go:112]: yes, add this for this case.

File: pkg/utils/utils.go (hunk 4 above)
Comment (zhiweiyin318) [reply pkg/utils/utils.go:37]: I found that we have false status for ManagedClusterConditionAvailable when managed cluster api server is Unavailable. https://github.com/open-cluster-management-io/registration/blob/e6435f32a3ed877b974f3299bd26db1101162f7b/pkg/spoke/managedcluster/healthcheck_controller.go#L96

Comment (zhiweiyin318) [reply pkg/utils/utils.go]: done

**Pull request merged**